### PR TITLE
feat: Shared-secret Registration

### PR DESCRIPTION
**Pull request by stoically**
Types for https://matrix-org.github.io/synapse/latest/admin_api/register_api.html

### CODE REVIEW

File: src/register_users/shared_secret_register/v1.rs [hunk 1]
```diff
@@ -0,0 +1,80 @@
...
+    metadata: {
+        description: "Shared-secret registration",
+        method: POST,
+        name: "shared_secret_register_v1",
+        unstable_path: "/_synapse/admin/v1/register",
+        rate_limited: false,
+        authentication: None,
+    }
+
+    #[derive(Serialize)]
```
Comment (jplatte) [inline src/register_users/shared_secret_register/v1.rs:16]: This isn't how requests are meant to be constructed. Ruma provides traits for (de)serializing requests and responses, and the `ruma_api!` macro implements those already, properly accounting for path fields, query parameters and setting the path.

Comment (stoically) [reply src/register_users/shared_secret_register/v1.rs:16]: How would the correct usage look like? And what kind of refactorings are you expecting for the PR?

Comment (jplatte) [reply src/register_users/shared_secret_register/v1.rs:16]: Correct usage would be to go through [`ruma::client::Client::send_request`](https://docs.rs/ruma/latest/ruma/client/struct.Client.html#method.send_request), `matrix_sdk::Client::send` or (lower level) [`OutgoingRequest::try_into_http_request`](https://docs.rs/ruma/latest/ruma/api/trait.OutgoingRequest.html#tymethod.try_into_http_request) for serialization plus [`IncomingResponse::try_from_http_response`](https://docs.rs/ruma/latest/ruma/api/trait.IncomingResponse.html#tymethod.try_from_http_response).

For this PR, I just want you to remove any `Serialize` and `Deserialize` derives on `request` / `response` blocks.

Comment (stoically) [reply src/register_users/shared_secret_register/v1.rs]: Thanks for the explanation! I was looking for the `OutgoingRequest` implementation in the locally generated docs, but probably just overlooked them.

Also adjusted the PR by removing de-/serialize.

**Review by jplatte — APPROVED** [review]

File: src/register_users/shared_secret_register_nonce/v1.rs [hunk 2]
```diff
@@ -0,0 +1,37 @@
+//! [GET /_synapse/admin/v1/register](https://matrix-org.github.io/synapse/latest/admin_api/register_api.html#shared-secret-registration)
+
+use ruma::api::ruma_api;
+use serde::Deserialize;
```
Comment (jplatte) [inline src/register_users/shared_secret_register_nonce/v1.rs:4]:
```suggestion
```

**Pull request merged**